### PR TITLE
EDTR: Collapsible Filter bar.

### PR DESCRIPTION
**Pull request by nicolad**
## Problem this Pull Request solves
Implements feature described here: https://github.com/eventespresso/event-espresso-core/issues/1771

## How has this been tested
Locally

## Checklist

* [x] I have read the documentation relating to systems affected by this pull request, see https://github.com/eventespresso/event-espresso-core/tree/master/docs
* [ ] User input is adequately validated and sanitized
* [ ] all publicly displayed strings are internationalized (usually using `esc_html__()`, see https://codex.wordpress.org/I18n_for_WordPress_Developers)
* [ ] My code is tested.
* [ ] My code follows the Event Espresso code style.
* [ ] My code has proper inline documentation.
* [ ] My code accounts for when the site is in Maintenance Mode (MM2 especially disallows usage of models)

### CODE REVIEW

**Comment by nicolad** [comment]
There are a few related minor improvements I would like to do here:
1. Based on how things are organized in `Media Library`, I think it would make sense to push search bar to the right and keep `perPageControl`, `listViewButton`,`gridViewButton` and `filter` at the left side.
image

**Comment by tn3rb** [comment]
oh hey I missed your comment about changing things to match the Media Library prior to doing my code review...

ya I like your suggestions to do that. As well, can we use the same icon for the list view that they are using ?

So i guess we would end up with the following button order:

`[list view]` `[grid view]` `[filters]` 

with the search bar on the far right, and then when the filters are open, it would display the rest of the filters.

**Comment by nicolad** [comment]
> oh hey I missed your comment about changing things to match the Media Library prior to doing my code review...
> 
> ya I like your suggestions to do that. As well, can we use the same icon for the list view that they are using ?
> 
> So i guess we would end up with the following button order:
> 
> `[list view]` `[grid view]` `[filters]`
> 
> with the search bar on the far right, and then when the filters are open, it would display the rest of the filters.
Ok, thx.
Is the `per page` button ok to be the first item like here?
Peek 2019-10-30 23-19

**Comment by nicolad** [comment]
> oh hey I missed your comment about changing things to match the Media Library prior to doing my code review...
> 
> ya I like your suggestions to do that. As well, can we use the same icon for the list view that they are using ?
> 
> So i guess we would end up with the following button order:
> 
> `[list view]` `[grid view]` `[filters]`
> 
> with the search bar on the far right, and then when the filters are open, it would display the rest of the filters.

Yes, will see how it would be possible to get those icons.

**Comment by tn3rb** [comment]
hmmm... i think we should move the `per page` into the other filter bar then because that will look far more clean. Doesn't look right as is.

Let's see what Josh thinks

**Comment by tn3rb** [comment]
> Is the search field one of the filtering options? Should it be in the collapsable filter bar too?

we're emulating the UI from the WP Media Library. Vadim posted a screenshot here:
https://github.com/eventespresso/event-espresso-core/pull/1774#issuecomment-548106758

**Comment by garthkoyle** [comment]
I guess it's not clear to me yet. In Vadim's screenshot of the WP media library, the search field is on the same level as the filters. Are those the same type of filters that we have hidden behind the filter icon? The search bar is not outside the filter layout.

**Comment by joshfeck** [comment]
> i think we should move the per page into the collapsable filter bar then because that will look far more clean. Doesn't look right as is. Maybe add it onto the far right of the other filters

That'd be fine to move the per page setting into the part that collapses away. The important thing is the pagination itself needs to be displayed in the default view if pagination links are present.

**Comment by tn3rb** [comment]
> In Vadim's screenshot of the WP media library, the search field is on the same level as the filters. Are those the same type of filters that we have hidden behind the filter icon?

It doesn't really matter what the "type" is, we can ultimately display them however we want, but the intent here is to change our filter layouts to be more inline with what WP core is doing.

- currently the "list view" and "grid view" buttons are on the far right, but WP core has theirs on the far left. So we are going to move ours.

- currently the search input is in the middle/left with the select inputs, but WP core has theirs on the far right. So we are going to move our search input to the far right as well.

- WP core has their other two filters immediately to the right of the "view" buttons but still on the far left of the entire filter bar, we have far more filters to display, so we are going to place the rest of ours into a collapsible container and then add a "show filters" button beside the "view" buttons. So this is a slight departure from what WP core is doing but that's because we have more additional filters than they do.

>  The search bar is not outside the filter layout.

I'm not sure what you mean by this. WP core has their search input on the far right and that is where we are going to put ours so that the layouts match and users don't have to go back n forth looking for things in the filter bar.

> The important thing is the pagination itself needs to be displayed in the default view if pagination links are present.

yes, regardless of whether the "per page" (or other filters) are being displayed, the pagination links will appear at the top of the entity lists. If a user wants to change the number of entities being displayed, they will have to click the "show filters" button and then change the value for the "per page" filter

**Comment by tn3rb** [comment]
@nicolad 

Just had a brief discussion with Garth regarding the filters and he would like us to also move the search input into the collapsible container.

So I think we are going to end up with something like this:

```jsx
<div className="ee-entity-list-filter-bar-wrapper">
	<div className="ee-filter-bar-filter-main ee-filter-bar-filter">
		{ listViewButton }
		{ gridViewButton }
		{ showFiltersButton }
	</div>
	<div className="ee-filter-bar-filter-visible/hidden ee-filter-bar-filter">
		{ entityFilters }
		{ perPageControl }
		{ searchInput }
	</div>
</div>
``` 

where `ee-filter-bar-filter-main` will be left aligned and always visible
and `ee-filter-bar-filter-visible/hidden` is the collapsible container where the css class name changes between `ee-filter-bar-filter-visible` and `ee-filter-bar-filter-hidden`

Feel free to modify the actual structure if you need to and ping me in chat if anything is not clear or you need any kind of help.

File: assets/src/hooks/useRect.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+				resizeObserver.disconnect();
+				resizeObserver = null;
+			};
+		}
+		// Browser support, remove freely
+		window.addEventListener( 'resize', handleResize );
+
+		return () => {
+			window.removeEventListener( 'resize', handleResize );
+		};
```
[the remark this answers is not in the document]
Comment (nicolad) [reply assets/src/hooks/useRect.js]: Yes, good suggestion, it improves readability. I had applied that.

**Comment by nicolad** [comment]
For the record, `useRect` hooks is reusing code from here: https://gist.github.com/morajabi/523d7a642d8c0a2f71fcfa0d8b3d2846
Was thinking about this one: https://ui.reach.tech/rect/
But this adds up another dependency, so I went with first option.

**Comment by nicolad** [comment]
@manzoorwanijk have created a separate issue to track that specific lag: https://github.com/eventespresso/event-espresso-core/issues/1808
***
We might have, in the meantime, other animations implemented, so will try to get a common optimization there.

**Pull request merged**